Optim foreach cleanup for RAdam

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* __->__ #70230
* #70229
* #69982
* #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to RAdam optimizer + cleanup

Differential Revision: [D33767874](https://our.internmc.facebook.com/intern/diff/D33767874)

mikaylagawarecki committed:
Optim foreach cleanup for RAdam

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/radam.py:
```diff
@@ -115,14 +121,140 @@ def step(self, closure=None):
+            if rho_t > 5 else 0 for rho_t in rho_t_list]
+    unrectified = [0 if rect > 0 else 1. for rect in rect]
+
+    exp_avg_sq_sqrt = torch._foreach_add(torch._foreach_sqrt(exp_avg_sqs), eps)
```
`eps` was unused in the multitensor radam implementation but used in the single tensor implentation, I believe the correct place to add it is here but I am not super sure whether this omission was intentional, should i be adding this here?

albanD replied:
I wouldn't change the foreach implementation at all in this PR and open an issue about this.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

albanD commented on torch/optim/radam.py:
```diff
@@ -115,14 +125,150 @@ def step(self, closure=None):
+          exp_avgs: List[Tensor],
+          exp_avg_sqs: List[Tensor],
+          state_steps: List[int],
+          foreach: bool = None,
```
nit torchscript comment.

albanD replied:
nit torchscript comment.

albanD reviewed (commented):
Looks good.
I think we want to avoid any change in behavior in this PR and move that to later.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

albanD reviewed (approved):
LGTM

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767874).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for RAdam"


Differential Revision: [D33767874](https://our.internmc.facebook.com/intern/diff/D33767874)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767874).
